File: FlutterMobileapp/lib/services/todo_service.dart
import '../core/api/api_client.dart';
import '../core/api/api_config.dart';
import '../models/todo.dart';

/// Todo Service
/// Handles all todo-related API operations
class TodoService {
  /// Get all todos with optional filters
  static Future<List<Todo>> getTodos({
    TodoStatus? status,
    TodoPriority? priority,
    int? limit,
  }) async {
    String endpoint = ApiConfig.todos;

    final params = <String, String>{};
    if (status != null) params['status'] = status.value;
    if (priority != null) params['priority'] = priority.value;
    if (limit != null) params['limit'] = limit.toString();

    if (params.isNotEmpty) {
      endpoint +=
          '?${params.entries.map((e) => '${e.key}=${e.value}').join('&')}';
    }

    final response = await ApiClient.get(endpoint);

    if (response.data is List) {
      return (response.data as List)
          .map((json) => Todo.fromJson(json))
          .toList();
    }

    return [];
  }

  /// Get a single todo by ID
  static Future<Todo?> getTodoById(String id) async {
    final response = await ApiClient.get('${ApiConfig.todos}/$id');

    if (response.data != null) {
      return Todo.fromJson(response.data);
    }

    return null;
  }

  /// Create a new todo
  static Future<Todo> createTodo({
    required String title,
    String? description,
    TodoPriority priority = TodoPriority.medium,
    DateTime? dueDate,
  }) async {
    final response = await ApiClient.post(
      ApiConfig.todos,
      body: {
        'title': title,
        if (description != null && description.isNotEmpty)
          'description': description,
        'priority': priority.value,
        if (dueDate != null) 'dueDate': dueDate.toIso8601String(),
      },
    );

    return Todo.fromJson(response.data);
  }

  /// Update a todo
  static Future<Todo> updateTodo(
    String id, {
    String? title,
    String? description,
    TodoStatus? status,
    TodoPriority? priority,
    DateTime? dueDate,
  }) async {
    final body = <String, dynamic>{};
    if (title != null) body['title'] = title;
    if (description != null) body['description'] = description;
    if (status != null) body['status'] = status.value;
    if (priority != null) body['priority'] = priority.value;
    if (dueDate != null) body['dueDate'] = dueDate.toIso8601String();

    final response = await ApiClient.put('${ApiConfig.todos}/$id', body: body);

    return Todo.fromJson(response.data);
  }

  /// Toggle todo status
  static Future<Todo> toggleTodoStatus(String id) async {
    final response = await ApiClient.patch('${ApiConfig.todos}/$id/toggle');

    return Todo.fromJson(response.data);
  }

  /// Delete a todo
  static Future<bool> deleteTodo(String id) async {
    await ApiClient.delete('${ApiConfig.todos}/$id');
    return true;
  }

  /// Get todo statistics
  static Future<TodoStats> getStats() async {
    final response = await ApiClient.get(ApiConfig.todoStats);

    return TodoStats.fromJson(response.data);
  }
}

/// Todo statistics
class TodoStats {
  final int total;
  final int pending;
  final int inProgress;
  final int completed;
  final int highPriority;

  TodoStats({
    required this.total,
    required this.pending,
    required this.inProgress,
    required this.completed,
    required this.highPriority,
  });

  factory TodoStats.fromJson(Map<String, dynamic> json) {
    return TodoStats(
      total: json['total'] ?? 0,
      pending: json['pending'] ?? 0,
      inProgress: json['in_progress'] ?? 0,
      completed: json['completed'] ?? 0,
      highPriority: json['high_priority'] ?? 0,
    );
  }
}

File: backend/src/services/todoService.js
/**
 * Todo Service
 * Business logic for todo operations
 */
const TodoModel = require('../models/Todo');
const logger = require('../config/logger');

/**
 * Create a new todo
 */
const createTodo = async (userId, todoData) => {
    logger.info('Creating new todo', { userId, title: todoData.title });

    const todo = await TodoModel.create({
        userId,
        title: todoData.title,
        description: todoData.description,
        priority: todoData.priority,
        dueDate: todoData.dueDate,
    });

    logger.info('Todo created successfully', { todoId: todo.id, userId });
    return todo;
};

/**
 * Get all todos for a user with optional filters
 */
const getTodos = async (userId, filters = {}) => {
    logger.info('Fetching todos', { userId, filters });

    const todos = await TodoModel.findByUserId(userId, {
        limit: filters.limit || 50,
        status: filters.status,
        priority: filters.priority,
        sortBy: filters.sortBy || 'createdAt',
        sortOrder: filters.sortOrder || 'desc',
    });

    logger.info('Todos fetched', { userId, count: todos.length });
    return todos;
};

/**
 * Get a single todo by ID
 */
const getTodoById = async (todoId, userId) => {
    logger.info('Fetching todo', { todoId, userId });

    const todo = await TodoModel.findById(todoId);

    if (!todo) {
        logger.warn('Todo not found', { todoId });
        return null;
    }

    // Verify ownership
    if (todo.userId !== userId) {
        logger.warn('Unauthorized todo access attempt', { todoId, userId });
        return null;
    }

    return todo;
};

/**
 * Update a todo
 */
const updateTodo = async (todoId, userId, updates) => {
    logger.info('Updating todo', { todoId, userId, updates: Object.keys(updates) });

    // Verify ownership first
    const existing = await getTodoById(todoId, userId);
    if (!existing) {
        return null;
    }

    const allowedUpdates = ['title', 'description', 'status', 'priority', 'dueDate'];
    const filteredUpdates = {};

    for (const key of allowedUpdates) {
        if (updates[key] !== undefined) {
            filteredUpdates[key] = updates[key];
        }
    }

    const updated = await TodoModel.update(todoId, filteredUpdates);
    logger.info('Todo updated successfully', { todoId });
    return updated;
};

/**
 * Toggle todo status
 */
const toggleTodoStatus = async (todoId, userId) => {
    logger.info('Toggling todo status', { todoId, userId });

    // Verify ownership first
    const existing = await getTodoById(todoId, userId);
    if (!existing) {
        return null;
    }

    const updated = await TodoModel.toggleStatus(todoId);
    logger.info('Todo status toggled', { todoId, newStatus: updated?.status });
    return updated;
};

/**
 * Delete a todo
 */
const deleteTodo = async (todoId, userId) => {
    logger.info('Deleting todo', { todoId, userId });

    // Verify ownership first
    const existing = await getTodoById(todoId, userId);
    if (!existing) {
        return false;
    }

    await TodoModel.delete(todoId);
    logger.info('Todo deleted successfully', { todoId });
    return true;
};

/**
 * Get todo statistics for a user
 */
const getStats = async (userId) => {
    logger.info('Fetching todo stats', { userId });
    return TodoModel.getStatsByUserId(userId);
};

module.exports = {
    createTodo,
    getTodos,
    getTodoById,
    updateTodo,
    toggleTodoStatus,
    deleteTodo,
    getStats,
};
